stdlib/asyncio/windows_utils.pyi

```diff
@@ -36,3 +36,31 @@ if sys.platform == "win32":
         def handle(self) -> int: ...
         def fileno(self) -> int: ...
         def close(self, *, CloseHandle: Callable[[int], None] = ...) -> None: ...
+
+    class Popen(subprocess.Popen[AnyStr]):
+        # The docstring says "The stdin, stdout, stderr are None or instances of PipeHandle."
+        # but that is incorrect: if you pass in a file object, that will still be
+        # put in the stdin/stdout/stderr attributes.
```

Comment:
Are you sure this is correct? The attributes are set to file objects in `super().__init__()`, but later replaced in the `else:` below.

Reply:
I might be misreading the code; I don't have access to Windows myself. But I think it sets these attrs to a PipeHandle only if stdin == PIPE, so if you pass stdin=<some file object>, that file object ends up in self.stdin.

Reply:
Looks like @Akuli is correct. If I apply this patch(!) to `asyncio/windows_utils.py` in my local clone of CPython:

```diff
diff --git a/Lib/asyncio/windows_utils.py b/Lib/asyncio/windows_utils.py
index ef277fac3e..24dda3263d 100644
--- a/Lib/asyncio/windows_utils.py
+++ b/Lib/asyncio/windows_utils.py
@@ -171,3 +171,4 @@ def __init__(self, args, stdin=None, stdout=None, stderr=None, **kwds):
                 os.close(stdout_wfd)
             if stderr == PIPE:
                 os.close(stderr_wfd)
+        print(f'{type(self.stdin)=}')
```

... then this is what I get in an interactive shell:

```python
Python 3.11.0a5+ (main, Feb 11 2022, 10:54:07) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> from asyncio.windows_utils import Popen
>>> with open('foo.txt', 'w') as f:
...     p = Popen(sys.executable, stdin=f)
...
type(self.stdin)=<class 'NoneType'>
```

Reply:
I'm starting to think that maybe you aren't supposed to pass in file objects at all, only `PipeHandle`s or `None` or `PIPE`.

For testing this on windows, maybe print each of `self.stdin`, `self.stdout` and `self.stderr`, and then run a subprocess with some high level function? As the file name suggests, this class seems to be mostly a platform-specific lower level implementation detail.

Reply:
@AlexWaygood Interesting, I'll have to look at the code to figure out how that's possible.

@Akuli Interestingly, `Popen` is not actually used or documented anywhere else in asyncio. It looks more like they kept it in by accident.

Reply:
We could always just cop out and use `Any`...

Reply:
It is used, at least in Python 3.9 where `windows_events.py` has this code:
```
class _WindowsSubprocessTransport(base_subprocess.BaseSubprocessTransport):

    def _start(self, args, shell, stdin, stdout, stderr, bufsize, **kwargs):
        self._proc = windows_utils.Popen(
            args, shell=shell, stdin=stdin, stdout=stdout, stderr=stderr,
            bufsize=bufsize, **kwargs)
```

Reply:
Yes, the use is still there in 3.11 https://github.com/python/cpython/blob/df9f7597559b6256924fcd3a1c3dc24cd5c5edaf/Lib/asyncio/windows_events.py#L890